### PR TITLE
Make logging for validation optional

### LINKED ISSUE #5794: Downgrade ERROR to something less less terrifying?
A pretty frequent source of confusion is the "errors" that appear in debug.log that really shouldn't be, is there any reason that these shouldn't be changed to warnings or just informational prints? "Error" just comes on a bit too strong for my liking and doesn't really reflect what the log print is talking about (rejected transaction, not someone hitting the self destruct button).

Propose:

    ERROR: AcceptToMemoryPool : non-final

Becomes:

    INFO: AcceptToMemoryPool : non-final

### PR DESCRIPTION
**Pull request by laanwj**
Avoid ERROR messages like `ERROR: AcceptToMemoryPool : non-final`. Logging of incoming transaction validation failures can be enabled optionally with`-debug=mempoolrej`.

Also introduces a consistent format for transaction validation failures that always includes the txid.

```
2015-08-05 13:25:26 txfail 4e5014c4084aab04aac280e4dc6e26ffbc76ea0feb817b98a1a67d8b18d4e80b: AcceptToMemoryPool: inputs already spent
```

I have opted for this approach instead of the one outlined in #5794 to remove all messages, because they provide extra troubleshooting information.

Fixes #5794.

### CODE REVIEW

**Comment by sipa** [comment]
Hmm, have you considered to not print the error immediately, but only report through ValidationState, and then have ProcessMessage choose (optionally) to report the error string it, if any?

That's also more compatible with future modularization of the consensus code, where you want to do the actual logging externally.

**Comment by laanwj** [comment]
Yes let's do that...

**Comment by laanwj** [comment]
Done. CValidationState is augmented with optional debug information, which is logged upstream where possible.

This makes the change larger than originally intended, but the end result is better.
See commit messages/details for more information.

N.B.: I've also removed the logging in `CScriptCheck::operator()()`. This is consistent but I'm uncertain about this, because this means that the script error is no longer logged in case of verification through CCheckQueue (so in the case of blocks). However it *is* logged when called directly from CheckInputs (https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1394).

File: src/main.cpp [hunk 1]
```diff
@@ -778,6 +769,14 @@ CAmount GetMinRelayFee(const CTransaction& tx, unsigned int nBytes, bool fAllowF
     return nMinFee;
 }
 
+/** Convert CValidationState to a human-readable message for logging */
+static std::string FormatStateMessage(const CValidationState &state)
+{
+    return strprintf("%s%s (code %i)",
+        state.GetRejectReason(),
+        state.GetDebugMessage().empty() ? "" : (", "+state.GetDebugMessage()),
```
Comment (sipa) [inline src/main.cpp]: No closing bracket?

**Comment by sipa** [comment]
@laanwj That's not very consistent, as I think that means it will be printed with -par=1 and not with other values. I don't know of a better solution, though.

**Comment by laanwj** [comment]
@sipa So we just call par=1 mode 'safe mode with debugging'. Voila.

File: src/main.cpp (hunk 1 above)
Comment (laanwj) [inline src/main.cpp]: The bracket is not in the string.

**Comment by sipa** [comment]
Untested ACK (but see nit about closing bracket).

File: src/main.cpp (hunk 1 above)
Comment (laanwj) [inline src/main.cpp]: Removed the parenthesis around ", "+... - they made the code harder to read instead of easier.

**Comment by dexX7** [comment]
> Also introduces a consistent format for transaction validation failures that always includes the txid.

Great PR, I'm running it for some hours, and this is really useful!

File: src/main.h [hunk 2]
```diff
@@ -455,7 +455,14 @@ extern CBlockTreeDB *pblocktree;
  */
 int GetSpendHeight(const CCoinsViewCache& inputs);
 
+/** Reject codes greater or equal to this can be returned by AcceptToMemPool
+ * for transactions, to signal internal conditions. They cannot and should not
+ * be sent over the P2P network.
+ */
+static const unsigned int REJECT_INTERNAL = 0x100;
 /** local "reject" message codes for RPC which can not be triggered by p2p trasactions */
```
Comment (morcos) [inline src/main.h]: This comment isn't accurate is it?

**Comment by morcos** [comment]
I really like this idea, there are going to be a whole slew of new possible reject reasons with 6470 and with limited mempool chains.  It will be nice to be able to output them in debugging situations.

**Comment by dcousens** [comment]
concept ACK

**Comment by skarra** [comment]
:+1:

File: src/main.h (hunk 2 above)
Comment (laanwj) [inline src/main.h]: Correct. This comment only holds for HIGHFEE.

File: src/main.h [hunk 3]
```diff
@@ -455,7 +455,16 @@ extern CBlockTreeDB *pblocktree;
  */
 int GetSpendHeight(const CCoinsViewCache& inputs);
 
-/** local "reject" message codes for RPC which can not be triggered by p2p trasactions */
+/** Reject codes greater or equal to this can be returned by AcceptToMemPool
+ * for transactions, to signal internal conditions. They cannot and should not
+ * be sent over the P2P network.
+ */
+static const unsigned int REJECT_INTERNAL = 0x100;
```
Comment (droark) [inline src/main.h]: Just curious, was the duplication of the REJECT_HIGHFEE value intentional?

**Comment by jtimon** [comment]
I've been trying to "move error messages up" for consensus and policy functions in different ways and with multiple attempts since at least January 15th 2015 (the first version of #5669 which also moved CheckTransaction to consensus/consensus did that).
I would have appreciated to be asked for review on this one. Now it's too late to complain about its flaws but I still will do it.

**Pull request merged**